esp-hal/src/rng/mod.rs

```diff
@@ -0,0 +1,434 @@
+//!
+//! ### Example
+//!
+//! The following example demonstrates how to set up a [custom backend] using `getrandom v0.3.3`:
+//!
+//! ```rust, ignore
+//! use getrandom::Error;
+//!
+//! #[no_mangle]
+//! unsafe extern "Rust" fn __getrandom_v03_custom(
+//!     dest: *mut u8,
+//!     len: usize,
+//! ) -> Result<(), Error> {
+//!     unsafe { esp_hal::rng:Rng::new().read_into_raw() };
+//!     Ok(())
+//! }
+//! ```
+//!
+//! [`getrandom`]: https://crates.io/crates/getrandom
+//! [custom backend]: https://github.com/rust-random/getrandom/tree/v0.3.3?tab=readme-ov-file#custom-backend
+//!
+//! ## Examples
+//!
+//! ### Basic RNG operation
+//!
+//! ```rust, no_run
+//! # {before_snippet}
+//! # use esp_hal::rng::Rng;
+//!
+//! let rng = Rng::new();
+//!
+//! // Generate a random word (u32):
+//! let rand_word = rng.random();
+//!
+//! // Fill a buffer with random bytes:
+//! let mut buf = [0u8; 16];
+//! rng.read(&mut buf);
+//!
+//! loop {}
+//! # }
+//! ```
+//!
+//! ### TRNG operation
+//! ```rust, no_run
+//! # {before_snippet}
+//! # use esp_hal::Blocking;
+//! # use esp_hal::rng::Trng;
+//! # use esp_hal::peripherals::Peripherals;
+//! # use esp_hal::peripherals::ADC1;
+//! # use esp_hal::analog::adc::{AdcConfig, Attenuation, Adc};
+//! #
+//! let mut buf = [0u8; 16];
+//!
+//! // ADC is not available from now
+//! let trng_source = TrngSource::new(peripherals.RNG, peripherals.ADC1.reborrow());
+//!
+//! let trng = Trng::try_new().unwrap(); // Unwrap is safe as we have enabled TrngSource.
+//!
+//! // Generate true random numbers
+//! trng.read(&mut buf);
+//! let true_random_number = trng.random();
+//!
+//! // Downgrade to Rng to allow disabling the TrngSource
+//! let rng = trng.downgrade();
+//!
+//! // Drop the true random number source. ADC is available now.
+//! core::mem::drop(trng_source);
+//!
+//! # {analog_pin}
+//!
+//! let mut adc1_config = AdcConfig::new();
+//! let mut adc1_pin = adc1_config.enable_pin(analog_pin, Attenuation::_11dB);
+//! let mut adc1 = Adc::<ADC1, Blocking>::new(peripherals.ADC1, adc1_config);
+//! let pin_value: u16 = nb::block!(adc1.read_oneshot(&mut adc1_pin))?;
+//!
+//! // Now we can only generate pseudo-random numbers...
+//! rng.read(&mut buf);
+//! let pseudo_random_number = rng.random();
+//!
+//! // ... but the ADC is available for use.
+//! let pin_value: u16 = nb::block!(adc1.read_oneshot(&mut adc1_pin))?;
+//! # {after_snippet}
+//! ```
+
+mod ll;
+
+use portable_atomic::{AtomicUsize, Ordering};
+
+static TRNG_ENABLED: AtomicUsize = AtomicUsize::new(0);
+static TRNG_USERS: AtomicUsize = AtomicUsize::new(0);
+
+use crate::peripherals::{ADC1, RNG};
+
+/// Ensures random numbers are cryptographically secure.
+pub struct TrngSource<'d> {
+    _rng: RNG<'d>,
+    _adc: ADC1<'d>,
+}
+
+impl<'d> TrngSource<'d> {
+    /// Enables the ADC entropy source.
+    // TODO: this is not final. A single ADC channel should be sufficient.
+    pub fn new(_rng: RNG<'d>, _adc: ADC1<'d>) -> Self {
```

Comment:
I suppose we can let `esp_wifi::init` also return `TrngSource<'d>` with the lifetime of `EspWifiController`, to avoid sacking (in this case the whole adc) an adc channel?